Be explicit about when VariantContexts are biallelic

davidbenjamin opened the pull request:
* VariantContexts from an assembled haplotype's EventMap, or found from pileups, that are really containers for a single alt allele, are explicitly marked as such.  This way we don't have to keep tracing back the source of a biallelic variant context and putting in little comments about why it's safe to assume it has only one alt allele.
* Some methods that remove or add haplotypes based on alleles found in pileups have been made void methods of AssemblyResultSet, which to mind mind is the appropriate way to encapsulate transformations acting on that class.
* other random simplification of code surrounding pileup haplotypes


@jamesemery This is a warmup PR for DRAGEN stuff, a bit of housekeeping of code at the margins of partially determined haplotype logic.

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/BiallelicVariant.java:
```diff
@@ -0,0 +1,33 @@
+    private static final long serialVersionUID = 1L;
+
+    public BiallelicVariant(final VariantContext vc) {
+        super(vc);
```
This is a good idea for a fix... I ALMOST want to suggest we disentangle this from all of the heavy VCF construction validation steps in HTSJDK that we usually undergo when we call the VCF constructor... those can actually show up as a non-insignificant amount of the runtime in profilers in some circumstances....

davidbenjamin replied:
I'm also tempted.  It's doable to replace this with a class (maybe call it `Event` to parallel `EventMap`?) that contains just a ref, alt, and position.  Think I should?

jamesemery replied:
I think in a lot of ways its cleaner... I like Event <-> EventMap... it could also have an event.asVCFLine() argument to convert it more lazily when we actually need a vcf in the code... It would mean pulling a lot of functionality that the htsjdk vcf was previously handing into this class however

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/EventMap.java:
```diff
@@ -15,72 +15,54 @@
 
     private static final Logger logger = LogManager.getLogger(EventMap.class);
-    protected static final int MIN_NUMBER_OF_EVENTS_TO_COMBINE_INTO_BLOCK_SUBSTITUTION = 3;
-    private static final int MAX_EVENTS_PER_HAPLOTYPE = 3;
```
Nice catch... these have not been used for quite a while...

jamesemery commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypingEngine.java:
```diff
@@ -124,7 +124,7 @@ public Config getConfiguration() {
      * @param vc                                 Input variant context to complete.
      * @return                                   VC with assigned genotypes
      */
-    public VariantContext calculateGenotypes(final VariantContext vc, final GenotypePriorCalculator gpc, final List<VariantContext> givenAlleles) {
```
Are BialellicVariantContainers ever making it here?

jamesemery commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java:
```diff
@@ -625,4 +629,175 @@ public void storeAssemblyHaplotypes() {
+
+        if (!haplotypesWithFilterAlleles.isEmpty()) {
+            if (argumentCollection.pileupDetectionArgs.debugPileupStdout) System.out.println("Found Assembly Haps with filtered Variants:\n"+haplotypesWithFilterAlleles.stream().map(Haplotype::toString).collect(Collectors.joining("\n")));
+            AssemblyBasedCallerUtils.haplotypeDeletionWarningLogger.warn(() -> "Haplotypes from Assembly are being filtered by heuristics from the PileupCaller. This can lead to missing variants. See --"+PileupDetectionArgumentCollection.PILEUP_DETECTION_FILTER_ASSEMBLY_HAPS_THRESHOLD+" for more details");
```
You might as well move the warning logger to this while you are moving things.

davidbenjamin replied:
done

jamesemery commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java:
```diff
@@ -473,13 +473,13 @@ static Set<Haplotype> filterPileupHaplotypes(final List<Haplotype> onlyNewHaplot
         // TODO this code might use some normalizing based on haplotype length in the future
         return onlyNewHaplotypes.stream()
-                .sorted(Comparator.comparingLong(hap -> -kmerizeSequence(hap.getBases(), kmerSize).stream()
+                .distinct() // TODO: The input haplotypes are not always unique.  Is this a horrible code smell?
```
though with representation issues i'm not surprised if you could accidentally re-create an existing haplotype by injecting a new allele into a bunch of the assembly haplotypes...

davidbenjamin replied:
Yeah, it's bad that this method is given a `List` with non-unique elements, but it's totally natural that the process of generating haplotypes would be redundant.  I'm going to use a `Set` instead of a `List` to deal with the problem upstream.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/BiallelicVariant.java:
```diff
@@ -0,0 +1,33 @@
+        return getAlternateAllele(0);
+    }
+
+    public boolean sameRefAndAlt(final BiallelicVariant other) {
```
done, and overrode hashcode as well

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/BiallelicVariant.java:
```diff
@@ -0,0 +1,33 @@
+ * Very simple class wrapping VariantContext when we want to be explicit that a variant is biallelic, such as
+ * in a haplotype EventMap
+ */
+public class BiallelicVariant extends VariantContext {
```
it doesn't extend VariantContext anymore

davidbenjamin commented:
#carrot(HaplotypeCaller CARROT Regression Tests, VariantCallingCarrotOrchestrated.gatk_docker, )

davidbenjamin commented:
@jamesemery The commit history is now so messy that the overall diff is much more intelligible than the individual commit diffs.  Few files have nontrivial changes.  Here's an overview:

- `AssemblyComplexity`, `TandemRepeat`, `VariantOverlapAnnotator`, `GenotypingEngine`,`AlleleFrequencyCalculator`, `AssemblyRegionTrimmer`, `PartiallyDeterminedHaplotypedComputationEngine`, `RampedHaplotypeCallerEngine`, `ReadThreadingAssembler`, `SomaticGenotypingEngine`, `HaplotypeBasedVariantRecaller`, `PartiallyDeterminedHaplotype`: the only real change in these classes is using `Event` instead of `VariantContext` in places where only a lightweight container for locus/ref/alt is required.  Some of the diffs are big but trivial.  (In `AssemblyRegionTrimmer` I also deleted the unused `nonVariantTargetRegions` method).
- `AlleleAndContext` and `LocationAndAlleles` were deleted as the new `Event` class makes them obsolete.
- `AlleleFiltering` is a big diff but it's just a bunch of replacing `AlleleAndContext` with `Event`.
- `HaplotypeCallerGenotypingEngine`: some trivial renaming and use of `Event` instead of `VariantContext`.
- `PileupDetectionArgumentCollection`: fixed a typo.
- `LeftAlignAndTrimVariants`: just some optimized imports.
- `Event`: this is the new class at the heart of the PR.
- `EventMap`: lighter than it was.  Storing a haplotype and other state were replaced by a static method to extract events from a haplotype.  Using `Event` instead of `VariantContext` allowed several things to be written more concisely, but no big-picture changes.  I didn't like how `buildEventMapForHaplotypes` both returned the set of event positions and filled in the haplotype's events as a side efffect, so I split it into a void method for filling events and another method for extracting start positions.
- `AlleleFilteringUnitTest`, `AssemblyBasedCallerUtilsUnitTest`, `HaplotypeCallerGenotypingEngineUnitTest.java`, `PartiallyDeterminedHaplotypeComputationEngineUnitTest`, `EventMapUnitTest`: everything that was tested before is still tested now.  Lots of replacing `VariantContext` with `Event`.  Deleted tests for a few unused methods here and there.
- The exact match expected VCF outputs in `HaplotypeCallerIntegrationTest`: no changes to the calls.  Slight changes in QUALs and QD annotation due to arbitrariness of selecting which pileup haplotypes to retain in case of ties in the read kmer support score.  I checked these all very carefully.  A bunch of VCF didn't change but their .idx files did when I turned on the update exact match toggle.

That's it for the trivially-changed classes.  Now for the few substantive changes. . .

davidbenjamin commented:
@jamesemery and now the overview of the more complex changes:

- `AssemblyResultSet`: the code for adding and removing haplotypes based on pileup alleles has become a `void` method of this class, where it belongs.  Here and elsewhere I introduce snappy variable and function named referring to "good" and "bad" alleles, which I find visually much clearer.  The code is basically the same as before but somewhat streamified.  I extracted a `makeHaplotypeWithInsertedEvent` method to eliminate some code duplication between GGA and pileup force-calling.
- `HaplotypeCallerEngine` and `Mutect2Engine`: Force-calling alleles are split into biallelic `Events`.  Duplicated code for finding all pileup events, then sifting them into good event to force-call and bad events to remove is extracted as `PileupBasedAlleles.goodAndBadPileupEvents`.  Computing `allVariationEvents` is much simpler because 1) it now uses `Event` instead of `VariantContext` and 2) `Event` overrides `equals` and `hashCode`.
- `PileupBasedAlleles`: `getPileupVariantContexts` and sorting into good and bad pileup variants has been unified into `goodAndBadPileupEvents()`.  It has additionally been somewhat rewritten for conciseness.  Also, instead of the somewhat kludgy method of making `VariantContext` with four temporary attributes, then filtering based on those attributes, it calculates the filtering status immediately and uses `Events`.  Also fixed the somewhat-misleading use of the word `alt` to mean `SNP`.
- `AssemblyBasedCallerUtils`: `applyPileupEventsAsForcedAlleles`, along with several helper methods that it calls, has been moved into `AssemblyResultResult`, where it is now a void member method.
- `GATKVariantContextUtils` mainly just using `Event` instead of `VariantContext`, which simplifies the code for splitting a `VariantContext` into biallelics.

After going through this exercise I realize that it's not actually so much.  The diff's bark is worse than its bite.  The overwhelming majority of changes are either replacing `VariantContext` with `Event` or moving methods to more appropriate classes.

davidbenjamin commented:
@jamesemery Back to you for further review.  This is big but not as monstrous as I made it out to be yesterday.

jamesemery commented:
I've started to take a look at this. Thank you for running the Carrot Tests! I guess you might be right about the score jitter not mattering but man there are enough differences in the HC integration test that i'm a little worried....

jamesemery commented on src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.WithIndels.gatk4.vcf:
```diff
@@ -95,31 +95,31 @@
-20	10074806	.	G	A	2832.06	.	AC=2;AF=1.00;AN=2;DP=75;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.38;QD=29.09;SOR=0.840	GT:AD:DP:GQ:PL	1/1:0,71:71:99:2846,213,0
-20	10075043	.	T	C	2420.06	.	AC=2;AF=1.00;AN=2;DP=64;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.64;QD=26.08;SOR=1.107	GT:AD:DP:GQ:PL	1/1:0,61:61:99:2434,184,0
-20	10075168	.	C	T	3627.06	.	AC=2;AF=1.00;AN=2;DP=91;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.37;QD=33.85;SOR=0.997	GT:AD:DP:GQ:PL	1/1:0,88:88:99:3641,264,0
+20	10074716	.	G	A	2415.06	.	AC=2;AF=1.00;AN=2;DP=70;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.67;QD=29.09;SOR=1.214	GT:AD:DP:GQ:PL	1/1:0,64:64:99:2429,192,0
```
I'm seeing some mess in the qual scores and here evidently all the QD scores specifically are broken... are you certain there isn't some ugly gotcha hiding in here? Do you have an explanation for exactly where the divergence from this branch is coming from?

davidbenjamin replied:
Remember the scary place in `AssemblyBasedCallerUtils` where the List of haplotypes to be sorted by their kmer support score had duplicates?  When I fixed the duplicate issue I broke all the tests because there are ties all over the place, so the exact (arbitrary) way in which you might modify the input order while filtering duplicates and sorting makes a difference.

The reason there are so many ties is that the score is extremely non-discriminating: as long as a kmer appears in even one read, it counts equally well, so the scoring often comes out like this: a few haplotypes with insertions leading, since longer haplotypes have more kmers with which to run up the score (I saw your comment about maybe normalzing by haplotype length -- your concern was well-placed), followed by a big tie among a lot of haplotypes with insertions.

I don't take broken tests lightly (well, to be honest I take it more lightly than, say, David Roazen but I do basically care about the integrity of the GATK) so I set some conditional breakpoints, and it turns out that the discrepancies only occur when the cutoff haplotype count occurred in the middle of a tie.  That is, the differences in QUAL and QD only appear when a tie must be broken arbitrarily.

After seeing this I decided to referee ties more permissively and keep everything with a score at least as good as the cutoff.  Thus, in a later commit of this PR, I decided that if the args are set to keep 5 haplotypes but the 4th through 8th have the same score we keep 8, for example.

jamesemery replied:
Okay thank you for this in-depth explanation. When I had written this comment out I had not yet internalized that these mismatches ONLY occur on the pileupcaller test files which is considerably lower impact than I had imagined when I saw the number of test files mismatching. That code in particular is less well tested/supported so our threshold for alarm is lower. The primary impact of this part of the codebase is relevant to DRAGEN-GATK which you are about to take ownership of anyway, so I fully expect anything that might be broken here to have a high chance of getting run to ground later in your process. I accept your explanation.... sorry to raise the alarm bells

davidbenjamin replied:
I double-checked this morning by reverting just the haplotype sorting and every DRAGEN test matches master.  It's less elegant because you have to introduce a `List`, then stream it and call `distinct()`, instead of using a `Set` from the beginning.

jamesemery commented on src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.testForceCallingNotProducingNoCalls.gatk4.vcf:
```diff
@@ -47,6 +47,7 @@
 20	10003832	.	G	A	2720.06	.	AC=2;AF=1.00;AN=2;DP=67;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=56.82;QD=29.03;SOR=1.473	GT:AD:DP:GQ:PL	1/1:0,66:66:99:2734,199,0
 20	10004094	.	A	C,T	1840.06	.	AC=2,0;AF=1.00,0.00;AN=2;DP=55;ExcessHet=0.0000;FS=0.000;MLEAC=2,0;MLEAF=1.00,0.00;MQ=52.62;QD=34.72;SOR=1.524	GT:AD:DP:GQ:PL	1/1:0,53,0:53:99:1854,157,0,1899,164,2091
 20	10004147	.	A	G	1852.06	.	AC=2;AF=1.00;AN=2;DP=54;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=55.19;QD=30.67;SOR=1.136	GT:AD:DP:GQ:PL	1/1:0,52:52:99:1866,156,0
+20	10004193	.	G	T	0	LowQual	AC=1;AF=0.500;AN=2;BaseQRankSum=-0.392;DP=54;ExcessHet=0.0000;FS=2.369;MLEAC=1;MLEAF=0.500;MQ=59.14;MQRankSum=3.385;QD=0.00;ReadPosRankSum=1.724;SOR=0.941	GT:AD:DP:GQ:PL	0/1:23,28:51:99:945,0,776
```
do you have a reason for this one? This seems to be more dangerous

davidbenjamin replied:
Just a difference in a marginal haplotype down the list being kept or not.  With a QUAL of 0 it's not going to affect the calls.

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/Event.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    // only use this for debugging, file output etc -- creating a full VariantContext is slow and defeats the purpose of this slim class
+    public VariantContext asVariantContext() {
```
add "@VisibleForTesting" and possibly make it package-protected

jamesemery commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java:
```diff
@@ -875,24 +601,23 @@ public static List<VariantContext> getVariantContextsFromGivenAlleles(final int
-                    }
-                });
-        return results;
+    public static List<VariantContext> getEventsFromActiveHaplotypes(final int loc, final List<Haplotype> haplotypes, final boolean includeSpanningEvents) {
```
this seems like a misnomer given that this is the exit point for returning VCs from the events...

davidbenjamin replied:
fixed

jamesemery commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java:
```diff
@@ -650,63 +456,28 @@ static Map<Kmer, Integer>  getKmerReadCounts(final List<GATKRead> hardClippedPil
-                        .collect(Collectors.toMap(
-                                Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+        // sort by score = # of kmers in haplotype that appear in any read
+        // TODO this code might use some normalizing based on haplotype length in the future
```
Honestly we should just make a ticket in github to revisit this entire codepath at some later date... i don't like this heuristic and i suspect we could do much better... thank you for not cutting too deeply here...

jamesemery replied:
also i will warn you this code WAS a performance hotspot so you might want to consider if any of these streams are going to be heavy duty.... Particularly anything looping over kmers...

davidbenjamin replied:
This shouldn't be a problem because it's only one stream per haplotype.  If it were streaming over every kmer (or of course if the CARROT tets showed a performance regression) I would be more worried.

jamesemery reviewed (commented):
Overall i think Event is much cleaner than the previous code and I quite like the changes you've made... I found one place where some unit tests changed because you changed the nature of a method test but mostly its suylistic/question comments and nothing too big. You are right this looked a lot worse than it was in reality...

jamesemery reviewed (commented):
Overall i think Event is much cleaner than the previous code and I quite like the changes you've made... I found one place where some unit tests changed because you changed the nature of a method test but mostly its suylistic/question comments and nothing too big. You are right this looked a lot worse than it was in reality...

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/Event.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    // only use this for debugging, file output etc -- creating a full VariantContext is slow and defeats the purpose of this slim class
+    public VariantContext asVariantContext() {
```
I eliminated the sourceless one and renamed the other `convertToVariantContext` in order to make clear that this is not cheap.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/Event.java:
```diff
@@ -0,0 +1,117 @@
+
+    public static Event of(final VariantContext vc) {
+        Utils.validateArg(vc.isBiallelic(), "variant must be biallelic");
+        return new Event(vc.getContig(), vc.getStart(), vc.getReference(), vc.getAlternateAllele(0));
```
I don't see any need in my code, but I renamed it to `ofWithoutAttributes` so people will know.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/TandemRepeat.java:
```diff
@@ -48,6 +49,12 @@ public Map<String, Object> annotate(final ReferenceContext ref,
         return Collections.unmodifiableMap(map);
     }
 
+    public static Pair<List<Integer>, byte[]> getNumTandemRepeatUnits(final ReferenceContext ref, final Event event) {
```
Yeah, I mean I could refactor the tandem repeat code but it would be hard to justify that as a contribution to DRAGEN GATK.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/readthreading/ReadThreadingAssembler.java:
```diff
@@ -917,7 +918,7 @@ public static void addAssembledVariantsToEventMapOutput(final AssemblyResultSet
                         queue.stream().limit(200).forEachOrdered(vc -> assembledEventMapVcfOutputWriter.get().add(vc));
                     }
-                    queue.add(event);}));
+                    queue.add(event.asVariantContext());}));
```
Switched the source to "assembly".

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/Event.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    // This should only be used once in the lifecycle of an event: when we make the jump from discovered event to variant context for output
+    public VariantContext asVariantContext(final String source) {
```
I don't quite feel passionate enough, especially after renaming it to something less lackadaisical, but I added some javadoc to the one place where the switch from `Event` to `VariantContext` happens.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:
```diff
@@ -2091,37 +2067,14 @@ public static int[] matchAllelesOnly(final VariantContext variant1, final Varian
-        return result;
+        // result[i] = j such that variant1's ith alt and variant2's jth alt are equivalent, or -1 if no such j exists
+        return IntStream.range(0, events1.size())
+                .map(i -> IntStream.range(0, events2.size()).filter(j -> events1.get(i).equals(events2.get(j))).findFirst().orElse(-1))
```
Matching only alleles without matching position is never used in GATK code.  I think I'll just rename the method to `matchAllelesAndStart`.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupBasedAlleles.java:
```diff
@@ -213,13 +168,12 @@ private static void validatePileupVariant(final VariantContext pileupVariant) {
      */
     @VisibleForTesting
-    static boolean evaluateBadRead(final GATKRead read, final ReferenceContext referenceContext, final PileupDetectionArgumentCollection args, final SAMFileHeader headerForRead) {
+    static boolean badPileupRead(final GATKRead read, final PileupDetectionArgumentCollection args, final SAMFileHeader headerForRead) {
```
Done, and likewise for `badAssemblyRead` -> `isBadAssemblyRead`

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupBasedAlleles.java:
```diff
@@ -52,9 +49,13 @@ public final class PileupBasedAlleles {
      * @param headerForReads                    Header for the reads (only necessary for SAM file conversion)
      * @return A list of variant context objects corresponding to potential variants that pass our heuristics.
      */
-    public static List<VariantContext> getPileupVariantContexts(final List<AlignmentAndReferenceContext> alignmentAndReferenceContextList, final PileupDetectionArgumentCollection args, final SAMFileHeader headerForReads, final int minBaseQualityScore) {
```
New javadoc.  I would like to "good" and "bad" because they make the code easier to read with the voice in one's head, but I took pains in the javadoc to be precise about the shorthand use of "good" and "bad".

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupBasedAlleles.java:
```diff
@@ -72,99 +73,73 @@ public static List<VariantContext> getPileupVariantContexts(final List<Alignment
+            Map<Byte, int[]> SNPCounts = new HashMap<>();
 
             for (PileupElement element : pileup) {
                 final byte eachBase = element.getBase();
```
Renamed it to `base`.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AssemblyComplexity.java:
```diff
@@ -177,12 +178,11 @@ private static boolean equalBasesExcludingSuffix(final byte[] eventMapBases, fin
-                .filter(vc -> !containsAltAllele(eventMap2, vc, 0))
+        return (int) hap1.getEventMap().getEvents().stream()
+                .filter(event -> event.getStart() != excludedPosition)
+                .filter(event -> !event.equals(eventMap2.get(event.getStart())))
```
It's equivalent, and in any case this is an M3 annotation so I would only be harming myself.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2Engine.java:
```diff
@@ -253,48 +253,34 @@ public List<VariantContext> callRegion(final AssemblyRegion originalAssemblyRegi
-            AssemblyBasedCallerUtils.applyPileupEventsAsForcedAlleles(originalAssemblyRegion, MTAC, aligner, assemblyResult.getReferenceHaplotype(), assemblyResult, pileupAllelesFoundShouldFilter, pileupAllelesPassingFilters, MTAC.pileupDetectionArgs.debugPileupStdout);
-        }
+        AssemblyResultSet assemblyResult = untrimmedAssemblyResult.trimTo(trimmingResult.getVariantRegion());
+        assemblyResult.addGivenAlleles(givenAlleles, MTAC.maxMnpDistance, aligner, MTAC.getHaplotypeToReferenceSWParameters());
```
Good point.  Switched the order.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/Event.java:
```diff
@@ -0,0 +1,117 @@
+
+    public boolean isSimpleDeletion() { return refAllele.length() > 1 && altAllele.length() == 1; }
+
+    public boolean isMNP() { return refAllele.length() > 1 && refAllele.length() == altAllele.length(); }
```
Absolutely -- any time `maxMnpDistance` is set to greater than 0 you can discover a MNP from a haplotype, hence ending with with a MNP in an `Event`.  It won't come up in DRAGEN, though.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyResultSet.java:
```diff
@@ -625,4 +629,162 @@ public void storeAssemblyHaplotypes() {
+    /**
+     * Remove haplotypes with alleles we wish to filter
+     * // TODO this is a bad algorithm for bad people -- it might eliminate good alleles on the same haplotypes
+     * // TODO: this should be a method of AssemblyResultSet, not an external static method
```
I took out the second TODO.  But as I understand your comment that it is a "bad algorithm for bad people" remains true.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/EventMap.java:
```diff
@@ -181,77 +144,45 @@ protected void processCigarForInitialEvents(final int maxMnpDistance) {
-        } else {
-            Utils.validateArg(!vc2.isSNP(), () -> "vc1 is " + vc1 + " but vc2 is a SNP, which implies there's been some terrible bug in the cigar " + vc2);
-        }
+    protected static Event combineEvents(final Event e1, final Event e2) {
```
Renamed it as `makeCompoundEvents` and expanded javadoc.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/EventMap.java:
```diff
@@ -15,104 +14,71 @@
+        return new EventMap(getEvents(haplotype, ref, haplotype.getLocation(), maxMnpDistance));
+    }
+
+    public static EventMap of(final Event ... events) {
```
VisibleForTesting for now.  I will ponder further as I get deeper into the PartiallyDeterminedHaplotype code.

davidbenjamin commented:
@jamesemery back to you.

jamesemery commented on src/main/java/org/broadinstitute/hellbender/utils/haplotype/Event.java:
```diff
@@ -114,4 +115,11 @@ private static boolean differentLastBase(final Allele ref, final Allele alt) {
+
+    @Override
+    public String toString() {
+        return String.format("[Event @ %s, %s->%s, %s",
```
I know its for debugging but the java String.format() call is MUCH slower than string.append() (which string (" " + " " + " ") maps into. These things get called enough in that debug path that I would consider keeping this fast.

davidbenjamin replied:
done

jamesemery reviewed (approved):
The refactor looks good to me. I would consider (mostly for your own sake) running this through the FE workspace one more time since there were non-zero changes to the dragen pipeline just to re-set your baseline. Otherwise this seems alright to merge (one comment)
